### PR TITLE
stm32: Configure ADC using new pinctrl via DT

### PR DESCRIPTION
**Pull request by martinjaeger**
- Add support for DT based pinmux configurations to the ADC driver.
- Fix ADC1/2 dts for STM32H7 board
- Use the new DT pinctrl property in all STM32 boards
- Remove old pinmux defines for ADC from `pinmux.c`

Checked with the script from @lochej that I didn't forget any boards.

### CODE REVIEW

File: dts/arm/st/h7/stm32h7.dtsi
Comment (lochej) [inline dts/arm/st/h7/stm32h7.dtsi:579]: That looks good, I check the RM for H7 series (RM0433 Rev7 page 1047)

image

File: boards/arm/nucleo_h743zi/nucleo_h743zi.dts
Comment (lochej) [inline boards/arm/nucleo_h743zi/nucleo_h743zi.dts:86]: I'm ok if you want to change the ADC definition other here, but please also update the index.rst for this board. I have this board so I'm going to try your PR when I have time today.

File: boards/arm/disco_l475_iot1/disco_l475_iot1.dts
Comment (lochej) [inline boards/arm/disco_l475_iot1/disco_l475_iot1.dts:220]:
```suggestion
	pinctrl-0 = <&adc1_in5_pa0 &adc1_in3_pc2
		     &adc1_in4_pc3 &adc1_in13_pc4
		     &adc1_in14_pc5>;
```
Little nit other here, I'd say 2 property on each lines is better readability

File: boards/arm/96b_aerocore2/96b_aerocore2.dts
Comment (lochej) [inline boards/arm/96b_aerocore2/96b_aerocore2.dts:125]:
```suggestion
&adc1 {
	pinctrl-0 = <&adc1_in10_pc0 &adc1_in11_pc1 
		     &adc1_in12_pc2 &adc1_in13_pc3>;
	status = "okay";
```

Thansk for your work, little nit for readability other here 😃

**Review by lochej — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_h743zi/nucleo_h743zi.dts
Comment (martinjaeger) [reply boards/arm/nucleo_h743zi/nucleo_h743zi.dts:86]: I actually think it was never the intention to use the special dual-mode ADC here. And I guess our driver doesn't support that properly either. That's why I changed to &adc1.

Updated the index.rst aswell. Would be nice if you could verify the PR, as I don't have this board. Thanks!

File: boards/arm/96b_aerocore2/96b_aerocore2.dts
Comment (lochej) [reply boards/arm/96b_aerocore2/96b_aerocore2.dts]: Thanks for updating !

File: boards/arm/nucleo_h743zi/nucleo_h743zi.dts
Comment (lochej) [reply boards/arm/nucleo_h743zi/nucleo_h743zi.dts:86]: Agreed, I will approve once I have tested with my board

Comment (erwango) [reply boards/arm/nucleo_h743zi/nucleo_h743zi.dts:86]: @gmarull Are you fie with this since you introduced this node ?

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
Approved but would need a rebase :-)

**Comment by martinjaeger** [comment]
> Approved but would need a rebase :-)

Yep, because the DAC was merged. Will rebase after @gmarull and @lochej have finished their reviews to reduce buildkite runs.

**Review by galak — APPROVED** [review]

**Review by lochej — APPROVED** [review]
LGTM, need rebase before merging and we're good to go 👍

File: boards/arm/nucleo_h743zi/nucleo_h743zi.dts
Comment (gmarull) [reply boards/arm/nucleo_h743zi/nucleo_h743zi.dts:86]: ADC on H7 used to work fine (using it on a custom board). I'll test it on my nucleo with and without these changes.

Comment (martinjaeger) [reply boards/arm/nucleo_h743zi/nucleo_h743zi.dts:86]: Ok, I have added a DNM. Can you please remove it once you're done with your test?

Comment (gmarull) [reply boards/arm/nucleo_h743zi/nucleo_h743zi.dts:86]: Ok, so it turns out `https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/drivers/adc/adc_api/src/test_adc.c#L201` should be `15` and not `0`. Once this is changed test seems to report coherent values. Should we fix it here or on a new PR? It's not related to the current changes.

**Review by gmarull — APPROVED** [review]

File: boards/arm/nucleo_h743zi/nucleo_h743zi.dts
Comment (martinjaeger) [reply boards/arm/nucleo_h743zi/nucleo_h743zi.dts:86]: Thanks for testing. I suggest to get this one merged first and have a dedicated PR for that issue. I actually think that the ADC test should be completely rewritten so that it obtains the used ADC channel from DTS instead of hard-coding it.

So I will remove the DNM again.

Comment (lochej) [reply boards/arm/nucleo_h743zi/nucleo_h743zi.dts:86]: > Ok, so it turns out `https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/drivers/adc/adc_api/src/test_adc.c#L201` should be `15` and not `0`. Once this is changed test seems to report coherent values. Should we fix it here or on a new PR? It's not related to the current changes.

Good catch, actually that was the problem ! 
I changed to 15 in test/adc/ac_api and using this PR:
```c
#elif defined(CONFIG_BOARD_NUCLEO_H743ZI)
#define ADC_DEVICE_NAME         DT_LABEL(DT_INST(0, st_stm32_adc))
#define ADC_RESOLUTION		16
#define ADC_GAIN		ADC_GAIN_1
#define ADC_REFERENCE		ADC_REF_INTERNAL
#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
#define ADC_1ST_CHANNEL_ID	15
```
Result with tying PA3 (A0) to VCC:
```
*** Booting Zephyr OS build zephyr-v2.4.0-550-g8ae325534e06  ***
Running test suite adc_basic_test
===================================================================
START - test_adc_sample_one_channel
Samples read: 0xffffffff 0xffff8000 0xffff8000 0xffff8000 0xffff8000 0xffff8000
 PASS - test_adc_sample_one_channel
===================================================================
START - test_adc_sample_two_channels
 SKIP - test_adc_sample_two_channels
===================================================================
START - test_adc_asynchronous_call
Samples read: 0xffffffff 0xffffffff 0xffffffff 0xffffffff 0xfffffffe 0xffff8000
 PASS - test_adc_asynchronous_call
...
```
This reads 0xFF ish as expected.

Tying to GND now:
```
*** Booting Zephyr OS build zephyr-v2.4.0-550-g8ae325534e06  ***
Running test suite adc_basic_test
===================================================================
START - test_adc_sample_one_channel
Samples read: 0x0000 0xffff8000 0xffff8000 0xffff8000 0xffff8000 0xffff8000
 PASS - test_adc_sample_one_channel
===================================================================
START - test_adc_sample_two_channels
 SKIP - test_adc_sample_two_channels
===================================================================
START - test_adc_asynchronous_call
Samples read: 0x0000 0x0000 0x0000 0x0000 0x0002 0xffff8000
 PASS - test_adc_asynchronous_call
===================================================================
...
```
This read 0x00 ish as expected ! 😄

Comment (lochej) [reply boards/arm/nucleo_h743zi/nucleo_h743zi.dts:86]: > Thanks for testing. I suggest to get this one merged first and have a dedicated PR for that issue. I actually think that the ADC test should be completely rewritten so that it obtains the used ADC channel from DTS instead of hard-coding it.


So now I fully approve this PR, all is working fine when the test actually uses the correct channel 🙃 . Agreed on fixing the ADC test in another PR.

**Pull request merged**
